KB: add multiple NICs issue (#26)
harvester/harvesterhci.io: kb/2022-03-10/multiple_nics_vm_connectivity.md

---
title: Multiple NICs VM Connectivity
descripion: How to deal VMs with multiple NICs in Harvester
slug: multiple-nics-vm-connectivity
authors:
  - name: Date Huang
    title: Software Engineer
    url: https://github.com/tjjh89017
    image_url: https://github.com/tjjh89017.png
tags: [vm, network]
hide_table_of_contents: false
---

# Multiple NICs VM Connectivity

## What will happen in VM with multiple NICs by default in Harvester

In some scenarios, we will setup two or more NICs for VM network to serve the different purpose. If setting up DHCP server and client by default in all networks, we might get connectivity issue randomly in VM. After some reboot, it might come back with expected connection, but it still will lost connection randomly after some period.

## Identify your connectivity issue that is covered in this knowledge base

> We take Linux VM and `iproute2` as example to identify your default route. If you cannot execute the commnad below, please switch root user or execute via `sudo` and execute again.

Author: ```suggestion
We take Linux VM and `iproute2` as example to identify your default route. If you cannot execute the commands below, please switch root user or execute via `sudo` and execute again.
```

> In your VM, execute following command in your VM in normal state:

Author: ```suggestion
In your VM, execute the following command in your VM in a normal state:
```

```bash
ip route show default
```

> In normal state, you will get only one route with destation to default and show the correct gateway IP.

Reviewer: destation --> destination

Author: ```suggestion
In the normal state, you will get only one route with destination to default and show the correct gateway IP.
```

```
default via <Gateway IP> dev eth0 proto dhcp src <VM IP> metric 100
```

For example:
```
$ ip route show default
default via 192.168.0.254 dev eth0 proto dhcp src 192.168.0.100 metric 100
$
```

> If issue that covered by this KB occurs, you could login VM via Web Console in Harvester Web UI and execute following command in your abnormal VM:

Reviewer: issue --> the issue

Author: ```suggestion
If the issue covered by this KB occurs, you could login VM via Web Console in Harvester Web UI and execute the following command in your abnormal VM:
```

```bash
ip route show default
```

In abnormal state, you will get one default route with incorrect gateway IP.
For example:
```
default via <Incorrect Gateway IP> dev eth0 proto dhcp src <VM's IP> metric 100
```

> ## Why cause connectivity issue randomly in VM.

Reviewer: Why the connectivity issue occurs randomly in VM

Author: ```suggestion
## Why the connectivity issue occurs randomly in VM
```

In most of time, we will enable cloud based VM with DHCP on all NICs and enable DHCP server with gateway IP. When DHCP client receives response and gets gateway IP, it will setup IP address and default route to the gateway IP, so users can access VM via SSH to the IP.

> Mostly Linux distribution will treat multiple NICs without any dependency and allow only one default route with same distance by default. It will start multiple DHCP clients simultaneously. If we have two NICs with DHCP and both have gateway IP in DHCP response, two DHCP client will race for setting default route to Linux network. According to different implementation of DHCP script, former and latter will not always win to setup default route. This will cause unstable default route in every DHCP renew process or machine boot up. VM's connection will be unstablei default route is always changing.

Reviewer: This will cause unstable default route in every DHCP renew process or machine boot up.
--> This will cause an unstable default route in every DHCP renewing process or machine boot up.

VM's connection will be unstablei default route is always changing.
--> VM's connection will be unstable if the default route is changing.

Author: ```suggestion
Mostly Linux distribution will treat multiple NICs without any dependency and allow only one default route with the same distance by default. It will start multiple DHCP clients simultaneously. If we have two NICs with DHCP and both have gateway IP in DHCP response, two DHCP clients will race for setting default route to Linux network. According to different implementations of DHCP script, the former and latter will not always win to set up the default route. This will cause an unstable default route in every DHCP renewing process or machine boot up. VM's connection will be unstable if the default route is changing.
```

## How to avoid random connectivity issue

We could avoid this issue easily by keeping VM only one NICs to serve network. We keep only one IP and one gateway in your system.

But not always we could use only one NIC to connect. For example, our infrasturcture will have storage network and service network. For security reason, storage network will be isolated from service network and have an independent subnet. In this situation, we must have two NICs to serve both service and storage network.

We could take one solution from below to fit your requirement and security policy.

### Disable DHCP, always setup Static IP on secondary NIC

The issue is caused by race condition of two DHCP clients. We could disable DHCP and always setup static IP in your VM setting. Or you could only set secondary NIC to static IP and keep primary one with DHCP enable.

> 1. edit `/etc/sysconfig/network/ifcfg-eth0` as primay NIC

Reviewer: Suggest adding the cloud-init example.

Author: add for static IP on secondary NIC.

## Identify your connectivity issue that is covered in this knowledge base

We take Linux VM and `iproute2` as example to identify your default route. If you cannot execute the commands below, please switch root user or execute via `sudo` and execute again.

In your VM, execute the following command in your VM in a normal state:
```bash
ip route show default
```

In the normal state, you will get only one route with destination to default and show the correct gateway IP.
```
default via <Gateway IP> dev eth0 proto dhcp src <VM IP> metric 100
```

For example:
```
$ ip route show default
default via 192.168.0.254 dev eth0 proto dhcp src 192.168.0.100 metric 100
$
```

If the issue covered by this KB occurs, you could login VM via Web Console in Harvester Web UI and execute the following command in your abnormal VM:
```bash
ip route show default
```

In abnormal state, you will get one default route with incorrect gateway IP.
For example:
```

> default via <Incorrect Gateway IP> dev eth0 proto dhcp src <VM's IP> metric 100

Author: it will depend on the expectation.
not sure how to generate an incorrect one

## How to avoid random connectivity issue

We could avoid this issue easily by keeping VM only one NICs to serve network. We keep only one IP and one gateway in your system.

But not always we could use only one NIC to connect. For example, our infrasturcture will have storage network and service network. For security reason, storage network will be isolated from service network and have an independent subnet. In this situation, we must have two NICs to serve both service and storage network.

We could take one solution from below to fit your requirement and security policy.

### Disable DHCP, always setup Static IP on secondary NIC

> The issue is caused by race condition of two DHCP clients. We could disable DHCP and always setup static IP in your VM setting. Or you could only set secondary NIC to static IP and keep primary one with DHCP enable.

Reviewer 2: ```suggestion
As explained above, the issue is caused by a "race condition" of two DHCP clients. One solution to avoid this issue, is to disable DHCP for the all NICs and configure them only with a static IP. In the same way, you could configure the secondary NIC with a static IP and keep the primary NIC with DHCP enabled.
```

1. edit `/etc/sysconfig/network/ifcfg-eth0` as primay NIC

```
BOOTPROTO='static'
IPADDR='192.168.0.100'
NETMASK='255.255.255.0'
```

> Or setup `eth0` with DHCP

Reviewer 2: ```suggestion
Or, if you want to keep the primary NIC with DHCP (`eth0` in this example), use the following values instead:
```

```
BOOTPROTO='dhcp'
DHCLIENT_SET_DEFAULT_ROUTE='yes'
```

2. edit route table `/etc/sysconfig/network/ifroute-eth0` (skipping if you used DHCP in eth0)

> Do not put other default route for your secondary NIC

Reviewer 2: ```suggestion
```

```
# Destination  Dummy/Gateway  Netmask  Interface
default        192.168.0.254  -        eth0
```

> 

Reviewer 2: ```suggestion

!!! warning
    Do not put other default route for your secondary NIC
    
```

> 3. setup secondary NIC, edit `/etc/sysconfig/network/ifcfg-eth1`

Reviewer 2: ```suggestion
3. Finally, configure the secondary NIC with a static IP by editing the file `/etc/sysconfig/network/ifcfg-eth1`:
```

```
BOOTPROTO='static'
IPADDR='10.0.0.100'
NETMASK='255.255.255.0'
```

> #### Cloud-Init config

Reviewer 2: Can we set the NIC names to the same as above or have a short introduction to the cloud-init file (when/where to use)

Author: enp1s0 or enp2s0 will be real-world naming in VM.
but the naming in Linux VM will not be always the same.
So I'm not sure.

if you prefer to replace with eth0/eth1 naming rule, I will follow your suggestion

Reviewer 2: that's what I thought, maybe just an intro to the cloud-init might be needed then (instead of renaming the interfaces).

Author: got it
let me change it to eth0/eth1

```yaml
network:
  version: 1
  config:
    - type: physical
      name: enp1s0
      subnets:
        - type: dhcp
    - type: physical
      name: enp2s0
      subnets:
        - type: static
          address: 10.0.0.100/24
```
   

> ### Disable second NIC default route from DHCP

Reviewer 2: ```suggestion
### Disable secondary NIC default route from DHCP
```

> If your NIC will need to get IP from DHCP, you should disable secondary NIC default route.

Reviewer 2: ```suggestion
If your secondary NIC requires to get its IP from DHCP, you'll need to disable the secondary NIC default route configuration.
```

> 1. edit `/etc/sysconfig/network/ifcfg-eth0` as primay NIC

Reviewer 2: ```suggestion
1. Confirm that the primary NIC configures its default route in the file `/etc/sysconfig/network/ifcfg-eth0`:
```

```
BOOTPROTO='dhcp'
DHCLIENT_SET_DEFAULT_ROUTE='yes'
```

> 2. setup secondary NIC, edit `/etc/sysconfig/network/ifcfg-eth1`

Reviewer 2: ```suggestion
2. Disable the secondary NIC default route configuration by editing the file `/etc/sysconfig/network/ifcfg-eth1`:
```

```
BOOTPROTO='dhcp'
DHCLIENT_SET_DEFAULT_ROUTE='no'
```

#### Cloud-Init config

> This solution is not available in Cloud-Init. Cloud-Init didn't allow any option for DHCP.

Reviewer 2: A cloud init section already exists above. Can we merge both?

Author: this cloud-init section is a sub-section of "Disable secondary NIC default route from DHCP".
It's not the same as the above one.
I want to mention there is no cloud-init config example and usage for "Disable secondary NIC default route from DHCP".

---

> title: Multiple NICs VM Connectivity

Reviewer 3: Probably use `VM Connectivity with Multiple NICs` is more preferred.

descripion: How to deal VMs with multiple NICs in Harvester
slug: multiple-nics-vm-connectivity
authors:
  - name: Date Huang
    title: Software Engineer
    url: https://github.com/tjjh89017
    image_url: https://github.com/tjjh89017.png
tags: [vm, network]
hide_table_of_contents: false
---

# Multiple NICs VM Connectivity

## What is the default behavior of a VM with multiple NICs

In some scenarios, you'll setup two or more NICs in your VM to serve different networking purposes. If all networks are setup by default with DHCP, you might get random connectivity issues. And while it might get fixed after rebooting the VM, it still will lose connection randomly after some period.

> 

Author: not sure how to link it will be better.

> 

Reviewer 3: how about [some scenarios](https://github.com/harvester/harvester/issues/1059)?

> 

Author: got it.

> 

Author: ```suggestion
In [some scenarios](https://github.com/harvester/harvester/issues/1059), you'll setup two or more NICs in your VM to serve different networking purposes. If all networks are setup by default with DHCP, you might get random connectivity issues. And while it might get fixed after rebooting the VM, it still will lose connection randomly after some period.

```

> 

Reviewer 3: seems this is not commited?

> 

Author: ok fixed it!

## How-to identify connectivity issues

In a Linux VM, you can use commands from the `iproute2` package to identify the default route.

In your VM, execute the following command:
```bash
ip route show default
```

> !!! note

Author: ```suggestion
!!! tip
```

    If you get the `access denied` error, please run the command using `sudo`
    

> The command's output will display only the `default` route with the gateway and VM IP of the main network interface (`eth0` in the example below).

Reviewer 3: ```suggestion
The output of this command will only show the default route with the gateway and VM IP of the primary network interface (`eth0` in the example below).
```

```
default via <Gateway IP> dev eth0 proto dhcp src <VM IP> metric 100
```

Here is the full example:
```
$ ip route show default
default via 192.168.0.254 dev eth0 proto dhcp src 192.168.0.100 metric 100
$

> ```

Reviewer 3: I thought we could just show the full example here because this is the basic stuff.

Author: but it will depend on different setup.
not easy to generate for a full example

> However, if the issue covered in this KB occurs, you'll only be able to connect to the VM via the Harvester Web Console.

Reviewer 3: ```suggestion
However, if the issue covered in this KB occurs, you'll only be able to connect to the VM via the VNC or serial console.
```

Once connected, you can run again the same command as before:
```bash

> ip route show default

Reviewer 3: append `$` upfront for good consistency.

```

However, this time you'll get a default route with an incorrect gateway IP.
For example:
```
default via <Incorrect Gateway IP> dev eth0 proto dhcp src <VM's IP> metric 100
```

> ## Why does connectivity issues occurs randomly

Author: I used this order because I want to check users' issues before we explain it to make sure their problem is covered by this KB.

> ## Why does connectivity issues occurs randomly

Reviewer 3: From the reader's point of view, I'm basically doing it by reading what is the problem is, why it occurs, and how to debug and fix it.

> ## Why does connectivity issues occurs randomly

Author: I will stay neutral on this.
but i recently read vmware kb.
take [this one](https://kb.vmware.com/s/article/67818?lang=en_US) as an example
they explained what happened before the reason.

> ## Why does connectivity issues occurs randomly

Reviewer 3: Yep, LGTM, it is just a suggestion instead of a requirement, thanks.

> ## Why does connectivity issues occurs randomly

Author: got it.
thanks a lot

In a standard setup, cloud based VMs are very often configured with DHCP for all NICs by default. The configuration sets an IP and a gateway for each NIC. Lastly, a default route to the gateway IP will also be added, so you can connect to the VM using its IP.

However, Linux distributions start multiple DHCP clients simultaneously and don't have a "prioritization" system. This means that in the case you have two or more NICs configured with DHCP, the clients will enter a "race condition" to configure the default route. And depending on the Linux distribution DHCP script currently running, there's no guarantee on which default route will be configured.

As the default route might change in every DHCP renewing process or after every OS reboot, this will create the network connectivity issues.

## How to avoid the random connectivity issues

> You could avoid these connectivity issues easily by having only one NIC attached to the VM and have only one IP and one gateway configured.

Reviewer 3: ```suggestion
You can easily avoid these connectivity issues by having only one NIC attached to the VM and having only one IP and one gateway configured.
```

However, the use of only one NIC is often not possible for VMs in a more complex infrastructure. For example, if your infrastructure has a storage network and a service network. For security reasons, the storage network will be isolated from the service network, and have an independent subnet. In this situation, you must have two NICs to connect to both the service and storage networks.

As a solution, you can choose one solution from below to fit your requirement and security policy.

### Disable DHCP on secondary NIC

> As explained above, the issue is caused by a "race condition" of two DHCP clients. One solution to avoid this issue, is to disable DHCP for the all NICs and configure them only with a static IP. In the same way, you could configure the secondary NIC with a static IP and keep the primary NIC with DHCP enabled.

Reviewer 3: ```suggestion
As mentioned above, the problem is caused by a `race condition` between two DHCP clients. One solution to avoid this problem is to disable DHCP for all NICs and configure them with static IPs only. Likewise, you can configure the secondary NIC with a static IP and keep the primary NIC enabled with DHCP.
```

1. To configure the primary NIC with a static IP (`eth0` in this example), you can edit the file `/etc/sysconfig/network/ifcfg-eth0` with the following values:

```
BOOTPROTO='static'
IPADDR='192.168.0.100'
NETMASK='255.255.255.0'
```

> Or, if you want to keep the primary NIC with DHCP (`eth0` in this example), use the following values instead:

Reviewer 3: ```suggestion
Alternatively, if you want to reserve the primary NIC using DHCP (`eth0` in this example), use the following values instead:
```

```
BOOTPROTO='dhcp'
DHCLIENT_SET_DEFAULT_ROUTE='yes'
```

> 2. You'll need to configure the default route by editing the file `/etc/sysconfig/network/ifroute-eth0` (skip this step if you configured your primary NIC with DHCP):

Reviewer 3: ```suggestion
2. You need to configure the default route by editing the file `/etc/sysconfig/network/ifroute-eth0` (if you configured the primary NIC using DHCP, skip this step):
```

```
# Destination  Dummy/Gateway  Netmask  Interface
default        192.168.0.254  -        eth0
```

!!! warning

>     Do not put other default route for your secondary NIC

Reviewer 3: ditto

Author: I will keep this as warning type.

Reviewer 3: sorry I mean the syntax here, if u run it with `yarn install && yarn start` u will have a preview of the blog, thanks.

Author: got it

> 3. Finally, configure the secondary NIC with a static IP by editing the file `/etc/sysconfig/network/ifcfg-eth1`:

Reviewer 3: ```suggestion
3. Finally, configure a static IP for the secondary NIC by editing the file `/etc/sysconfig/network/ifcfg-eth1`:
4. ```

```
BOOTPROTO='static'
IPADDR='10.0.0.100'
NETMASK='255.255.255.0'
```

#### Cloud-Init config

```yaml
network:
  version: 1
  config:
    - type: physical

>       name: enp1s0

Author: ```suggestion
      name: eth0
```

      subnets:
        - type: dhcp
    - type: physical

>       name: enp2s0

Author: ```suggestion
      name: eth1
```

tags: [vm, network]
hide_table_of_contents: false
---

# Multiple NICs VM Connectivity

## What is the default behavior of a VM with multiple NICs

In [some scenarios](https://github.com/harvester/harvester/issues/1059), you'll setup two or more NICs in your VM to serve different networking purposes. If all networks are setup by default with DHCP, you might get random connectivity issues. And while it might get fixed after rebooting the VM, it still will lose connection randomly after some period.

## How-to identify connectivity issues

In a Linux VM, you can use commands from the `iproute2` package to identify the default route.

In your VM, execute the following command:
```bash
ip route show default
```
:::tip
If you get the `access denied` error, please run the command using `sudo`
:::
    
The output of this command will only show the default route with the gateway and VM IP of the primary network interface (`eth0` in the example below).
```
default via <Gateway IP> dev eth0 proto dhcp src <VM IP> metric 100
```

Here is the full example:
```
$ ip route show default
default via 192.168.0.254 dev eth0 proto dhcp src 192.168.0.100 metric 100

> $

Author: ok fixed it!
